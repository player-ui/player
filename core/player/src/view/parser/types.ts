import type { Asset as AssetType, Expression, Binding } from '@player-ui/types';

export type AnyAssetType = AssetType<string>;
export enum NodeType {
  Asset = 'asset',
  View = 'view',
  Applicability = 'applicability',
  Template = 'template',
  Value = 'value',
  MultiNode = 'multi-node',
  Switch = 'switch',
  Async = 'async',
  Unknown = 'unknown',
  Empty = 'empty',
}
export declare namespace Node {
  export type ChildrenTypes = NodeType.Asset | NodeType.Value | NodeType.View;

  export interface Base<T extends NodeType> {
    /** Every node contains a type to distinguish it from other nodes */
    type: T;

    /** Every node (outside of the root) contains a reference to it's parent */
    parent?: Node;
  }

  export type PathSegment = string | number;

  export interface Child {
    /** The path of the child relative to the parent */
    path: PathSegment[];

    /** If true, the path points to an array, and the value will be appended to it result */
    array?: boolean;

    /** The child node */
    value: Node;
  }

  export interface BaseWithChildren<T extends NodeType> extends Base<T> {
    /** Any node that contains a list of children underneath it */
    children?: Child[];
  }

  export interface Asset<T extends AnyAssetType = AnyAssetType>
    extends BaseWithChildren<NodeType.Asset>,
      PluginOptions {
    /** Any asset nested within a view */
    value: T;
  }

  export interface View<T extends AnyAssetType = AnyAssetType>
    extends BaseWithChildren<NodeType.View>,
      PluginOptions {
    /** The root of the parsed view */
    value: T;
  }

  export interface Applicability extends Base<NodeType.Applicability> {
    /** The expression to execute that determines applicability of the target node */
    expression: Expression;

    /** The node to use if the expression is truthy */
    value: Node;
  }

  export interface Template extends Base<NodeType.Template> {
    /** The location of an array in the model */
    data: Binding;

    /** The template to use when mapping over the data */
    template: unknown;

    /** The number of nested templates so far */
    depth: number;

    /** should the template recomputed when data changes */
    dynamic?: boolean;
  }

  export interface Value
    extends BaseWithChildren<NodeType.Value>,
      PluginOptions {
    /** A simple node representing a value */
    value: any;
  }

  export interface MultiNode extends Base<NodeType.MultiNode> {
    /**
     * Should this list override the target node if they overlap?
     * If not amend the existing list
     */
    override?: boolean;

    /** A list of values that comprise this node */
    values: Array<Node>;
  }

  export interface Switch extends Base<NodeType.Switch> {
    /** Should this list be re-computed when data changes */
    dynamic?: boolean;

    /** A list of cases to evaluate in order */
    cases: SwitchCase[];
  }

  export interface SwitchCase {
    /** The expression to evaluate for a single case statement */
    case: Expression | true;
    /** The value to use if this case is true */
    value: Value;
  }

  export interface Async extends Base<NodeType.Async> {
    /** The unique id of the node */
    id: string;
    /** The value representing the node */
    value: Node;
  }

  export interface PluginOptions {
    /** A list of plugins */
    plugins?: {
      /** StringResolverPlugin options */
      stringResolver?: {
        /**
         * An optional array of node properties to skip during string resolution
         * Specified in the AssetTransformPlugin
         */
        propertiesToSkip?: string[];
      };
    };
  }

  export type Unknown = Base<NodeType.Unknown>;
  export type Empty = Base<NodeType.Empty>;
  export type ViewOrAsset = View | Asset;

  export type Node =
    | Asset
    | Applicability
    | Template
    | Value
    | View
    | MultiNode
    | Switch
    | Async
    | Unknown
    | Empty;
}
