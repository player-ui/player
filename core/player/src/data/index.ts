export * from './dependency-tracker';
export * from './model';
export * from './noop-model';
export * from './local-model';
