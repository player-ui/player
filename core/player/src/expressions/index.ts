export * from "./evaluator";
export * from "./types";
export * from "./utils";
export * from "./parser";
