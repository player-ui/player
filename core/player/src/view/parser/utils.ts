import type { Node } from './types';

/** Check to see if the object contains async */
export function hasAsync(obj: object): boolean {
  return Object.prototype.hasOwnProperty.call(obj, 'async');
}

/** Get the ID of the Node if there is one */
export function getNodeID(node?: Node.Node | null): string | undefined {
  if (!node) {
    return;
  }

  if (
    'value' in node &&
    typeof node.value === 'object' &&
    typeof node.value?.id === 'string'
  ) {
    return node.value.id;
  }
}
