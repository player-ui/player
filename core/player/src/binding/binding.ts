import { getBindingSegments } from "./utils";

export interface BindingParserOptions {
  /** Get the value for a specific binding */
  get: (binding: BindingInstance) => any;

  /**
   * Set the values for bindings.
   * This is used when the query syntax needs to modify an object
   */
  set: (transaction: Array<[BindingInstance, any]>) => void;

  /**
   * Get the result of evaluating an expression
   */
  evaluate: (exp: string) => any;

  /**
   * Without readOnly, if a binding such as this is used: arr[key='does not exist'],
   * then an object with that key will be created.
   * This is done to make assignment such as arr[key='abc'].val = 'foo' work smoothly.
   * Setting readOnly to true will prevent this behavior, avoiding unintended data changes.
   */
  readOnly?: boolean;
}

export type Getter = (path: BindingInstance) => any;

export type RawBindingSegment = number | string;
export type RawBinding = string | RawBindingSegment[];
export type BindingLike = RawBinding | BindingInstance;
export type BindingFactory = (
  raw: RawBinding,
  options?: Partial<BindingParserOptions>,
) => BindingInstance;

/**
 * A path in the data model
 */
export class BindingInstance {
  private split: RawBindingSegment[];
  private joined: string;
  private factory: BindingFactory;

  constructor(
    raw: RawBinding,
    factory = (rawBinding: RawBinding) => new BindingInstance(rawBinding),
  ) {
    const split = Array.isArray(raw) ? raw : raw.split(".");
    this.split = split.map((segment) => {
      if (typeof segment === "number") {
        return segment;
      }

      const tryNum = Number(segment);
      return isNaN(tryNum) ? segment : tryNum;
    });
    Object.freeze(this.split);
    this.joined = this.split.join(".");
    this.factory = factory;
  }

  asArray(): RawBindingSegment[] {
    return this.split;
  }

  asString(): string {
    return this.joined;
  }

  /**
   * Check to see if the given binding is a sub-path of the current one
   */
  contains(binding: BindingInstance): boolean {
    // need to account for partial key matches
    // [foo, bar] !== [foo, ba]
    const bindingAsArray = binding.asArray();

    if (bindingAsArray.length < this.split.length) {
      return false;
    }

    // Check every overlapping index to make sure they're the same
    // Intentionally use a for loop for speeeed
    for (let i = 0; i < this.split.length; i++) {
      if (this.split[i] !== bindingAsArray[i]) {
        return false;
      }
    }

    return true;
  }

  relative(binding: BindingInstance): RawBindingSegment[] {
    return this.asArray().slice(binding.asArray().length);
  }

  parent(): BindingInstance {
    return this.factory(this.split.slice(0, -1));
  }

  key(): RawBindingSegment {
    return this.split[this.split.length - 1];
  }

  /**
   * This is a utility method to get a binding that is a descendent of this binding
   *
   * @param relative - The relative path to descend to
   */
  descendent(relative: BindingLike): BindingInstance {
    const descendentSegments = getBindingSegments(relative);

    return this.factory(this.split.concat(descendentSegments));
  }
}
