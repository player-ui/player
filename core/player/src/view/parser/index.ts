import { omit, setIn } from 'timm';
import { SyncBailHook, SyncWaterfallHook } from 'tapable-ts';
import type { Template } from '@player-ui/types';
import type { AnyAssetType, Node } from './types';
import { NodeType } from './types';
import { getNodeID, hasAsync } from './utils';

export * from './types';
export * from './utils';

export const EMPTY_NODE: Node.Empty = {
  type: NodeType.Empty,
};

export interface ParseObjectOptions {
  /** how nested the templated is */
  templateDepth?: number;
}

interface NestedObj {
  /** The values of a nested local object */
  children: Node.Child[];

  value: any;
}
/**
 * The Parser is the way to take an incoming view from the user and parse it into an AST.
 * It provides a few ways to interact with the parsing, including mutating an object before and after creation of an AST node
 */
export class Parser {
  public readonly hooks = {
    /**
     * A hook to interact with an object _before_ parsing it into an AST
     *
     * @param value - The object we're are about to parse
     * @returns - A new value to parse.
     *  If undefined, the original value is used.
     *  If null, we stop parsing this node.
     */
    onParseObject: new SyncWaterfallHook<[object, NodeType]>(),

    /**
     * A callback to interact with an AST _after_ we parse it into the AST
     *
     * @param value - The object we parsed
     * @param node - The AST node we generated
     * @returns - A new AST node to use
     *   If undefined, the original value is used.
     *   If null, we ignore this node all together
     */
    onCreateASTNode: new SyncWaterfallHook<
      [Node.Node | undefined | null, object]
    >(),

    determineNodeType: new SyncBailHook<[object | string], NodeType>(),

    parseNode: new SyncBailHook<
      [
        obj: object,
        nodeType: Node.ChildrenTypes,
        parseOptions: ParseObjectOptions,
        determinedNodeType: NodeType | null
      ],
      Node.Node
    >(),
  };

  public parseView(value: AnyAssetType): Node.View {
    const viewNode = this.parseObject(value, NodeType.View);

    if (!viewNode) {
      throw new Error('Unable to parse object into a view');
    }

    return viewNode as Node.View;
  }

  private parseAsync(
    obj: object,
    type: Node.ChildrenTypes,
    options: ParseObjectOptions
  ): Node.Node | null {
    const parsedAsync = this.parseObject(omit(obj, 'async'), type, options);
    const parsedNodeId = getNodeID(parsedAsync);
    if (parsedAsync !== null && parsedNodeId) {
      return this.createASTNode(
        {
          id: parsedNodeId,
          type: NodeType.Async,
          value: parsedAsync,
        },
        obj
      );
    }

    return null;
  }

  public createASTNode(node: Node.Node | null, value: any): Node.Node | null {
    const tapped = this.hooks.onCreateASTNode.call(node, value);

    if (tapped === undefined) {
      return node;
    }

    return tapped;
  }

  /**
   * Checks if there are templated values in the object
   *
   * @param obj - The Parsed Object to check to see if we have a template array type for
   * @param localKey - The key being checked
   */
  private hasTemplateValues(obj: any, localKey: string) {
    return (
      Object.hasOwnProperty.call(obj, 'template') &&
      Array.isArray(obj?.template) &&
      obj.template.length &&
      obj.template.find((tmpl: any) => tmpl.output === localKey)
    );
  }

  public parseObject(
    obj: object,
    type: Node.ChildrenTypes = NodeType.Value,
    options: ParseObjectOptions = { templateDepth: 0 }
  ): Node.Node | null {
    const nodeType = this.hooks.determineNodeType.call(obj);

    if (nodeType !== undefined) {
      const parsedNode = this.hooks.parseNode.call(
        obj,
        type,
        options,
        nodeType
      );
      if (parsedNode) {
        return parsedNode;
      }
    }

    const parseLocalObject = (
      currentValue: any,
      objToParse: unknown,
      path: string[] = []
    ): NestedObj => {
      if (typeof objToParse !== 'object' || objToParse === null) {
        return { value: objToParse, children: [] };
      }

      const localObj = this.hooks.onParseObject.call(objToParse, type);

      if (!localObj) {
        return currentValue;
      }

      const objEntries = Array.isArray(localObj)
        ? localObj.map((v, i) => [i, v])
        : [
            ...Object.entries(localObj),
            ...Object.getOwnPropertySymbols(localObj).map((s) => [
              s,
              (localObj as any)[s],
            ]),
          ];

      const defaultValue: NestedObj = {
        children: [],
        value: currentValue,
      };

      const newValue = objEntries.reduce((accumulation, current): NestedObj => {
        const { children, ...rest } = accumulation;
        const [localKey, localValue] = current;
        if (localKey === 'asset' && typeof localValue === 'object') {
          const assetAST = this.parseObject(
            localValue,
            NodeType.Asset,
            options
          );

          if (assetAST) {
            return {
              ...rest,
              children: [
                ...children,
                {
                  path: [...path, 'asset'],
                  value: assetAST,
                },
              ],
            };
          }
        } else if (
          this.hooks.determineNodeType.call(localKey) === NodeType.Template &&
          Array.isArray(localValue)
        ) {
          const templateChildren = localValue
            .map((template: Template) => {
              const templateAST = this.hooks.onCreateASTNode.call(
                {
                  type: NodeType.Template,
                  depth: options.templateDepth ?? 0,
                  data: template.data,
                  template: template.value,
                  dynamic: template.dynamic ?? false,
                },
                template
              );

              if (templateAST?.type === NodeType.MultiNode) {
                templateAST.values.forEach((v) => {
                  // eslint-disable-next-line no-param-reassign
                  v.parent = templateAST;
                });
              }

              if (templateAST) {
                return {
                  path: [...path, template.output],
                  value: templateAST,
                };
              }

              // eslint-disable-next-line no-useless-return
              return;
            })
            .filter((element) => !!element);

          return {
            ...rest,
            children: [...children, ...templateChildren],
          } as NestedObj;
        } else if (
          localValue &&
          this.hooks.determineNodeType.call(localValue) === NodeType.Switch
        ) {
          const localSwitch = this.hooks.parseNode.call(
            localValue,
            NodeType.Value,
            options,
            NodeType.Switch
          );

          if (
            localSwitch &&
            localSwitch.type === NodeType.Value &&
            localSwitch.children?.length === 1 &&
            localSwitch.value === undefined
          ) {
            const firstChild = localSwitch.children[0];

            return {
              ...rest,
              children: [
                ...children,
                {
                  path: [...path, localKey, ...firstChild.path],
                  value: firstChild.value,
                },
              ],
            };
          }

          if (localSwitch) {
            return {
              ...rest,
              children: [
                ...children,
                {
                  path: [...path, localKey],
                  value: localSwitch,
                },
              ],
            };
          }
        } else if (localValue && hasAsync(localValue)) {
          const localAsync = this.parseAsync(
            localValue,
            NodeType.Value,
            options
          );
          if (localAsync) {
            children.push({
              path: [...path, localKey],
              value: localAsync,
            });
          }
        } else if (localValue && Array.isArray(localValue)) {
          const childValues = localValue
            .map((childVal) =>
              this.parseObject(childVal, NodeType.Value, options)
            )
            .filter((child): child is Node.Node => !!child);

          if (childValues.length > 0) {
            const multiNode = this.hooks.onCreateASTNode.call(
              {
                type: NodeType.MultiNode,
                override: !this.hasTemplateValues(localObj, localKey),
                values: childValues,
              },
              localValue
            );

            if (multiNode?.type === NodeType.MultiNode) {
              multiNode.values.forEach((v) => {
                // eslint-disable-next-line no-param-reassign
                v.parent = multiNode;
              });
            }

            if (multiNode) {
              return {
                ...rest,
                children: [
                  ...children,
                  {
                    path: [...path, localKey],
                    value: multiNode,
                  },
                ],
              };
            }
          }
        } else if (localValue && typeof localValue === 'object') {
          const determineNodeType =
            this.hooks.determineNodeType.call(localValue);

          if (determineNodeType === NodeType.Applicability) {
            const parsedNode = this.hooks.parseNode.call(
              localValue,
              NodeType.Value,
              options,
              determineNodeType
            );
            if (parsedNode) {
              return {
                ...rest,
                children: [
                  ...children,
                  {
                    path: [...path, localKey],
                    value: parsedNode,
                  },
                ],
              };
            }
          } else {
            const result = parseLocalObject(accumulation.value, localValue, [
              ...path,
              localKey,
            ]);
            return {
              value: result.value,
              children: [...children, ...result.children],
            };
          }
        } else {
          const value = setIn(
            accumulation.value,
            [...path, localKey],
            localValue
          );

          return {
            children,
            value,
          };
        }

        return accumulation;
      }, defaultValue);

      return newValue;
    };

    const { value, children } = parseLocalObject(undefined, obj);

    const baseAst =
      value === undefined && children.length === 0
        ? undefined
        : {
            type,
            value,
          };

    if (baseAst !== undefined && children.length > 0) {
      const parent = baseAst as Node.BaseWithChildren<any>;
      parent.children = children;
      children.forEach((child) => {
        // eslint-disable-next-line no-param-reassign
        child.value.parent = parent;
      });
    }

    return this.hooks.onCreateASTNode.call(baseAst, obj) ?? null;
  }
}
