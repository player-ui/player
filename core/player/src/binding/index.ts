import { SyncBailHook, SyncWaterfallHook } from 'tapable-ts';
import NestedError from 'nested-error-stacks';
import type { ParserResult, AnyNode } from '../binding-grammar';
import {
  // We can swap this with whichever parser we want to use
  parseCustom as parseBinding,
} from '../binding-grammar';
import type { BindingParserOptions, BindingLike } from './binding';
import { BindingInstance } from './binding';
import { isBinding } from './utils';
import type { NormalizedResult, ResolveBindingASTOptions } from './resolver';
import { resolveBindingAST } from './resolver';

export * from './utils';
export * from './binding';

export const SIMPLE_BINDING_REGEX = /^[\w\-@]+(\.[\w\-@]+)*$/;

const DEFAULT_OPTIONS: BindingParserOptions = {
  get: () => {
    throw new Error('Not Implemented');
  },
  set: () => {
    throw new Error('Not Implemented');
  },
  evaluate: () => {
    throw new Error('Not Implemented');
  },
};

/** A parser for creating bindings from a string */
export class BindingParser {
  private cache: Record<string, BindingInstance>;
  private parseCache: Record<string, ParserResult>;
  private parserOptions: BindingParserOptions;

  public hooks = {
    skipOptimization: new SyncBailHook<[string], boolean>(),
    beforeResolveNode: new SyncWaterfallHook<
      [AnyNode, Required<NormalizedResult> & ResolveBindingASTOptions]
    >(),
  };

  constructor(options?: Partial<BindingParserOptions>) {
    this.parserOptions = { ...DEFAULT_OPTIONS, ...options };
    this.cache = {};
    this.parseCache = {};
    this.parse = this.parse.bind(this);
  }

  /**
   * Takes a binding path, parses it, and returns an equivalent, normalized
   * representation of that path.
   */
  private normalizePath(
    path: string,
    resolveOptions: ResolveBindingASTOptions
  ) {
    if (
      path.match(SIMPLE_BINDING_REGEX) &&
      this.hooks.skipOptimization.call(path) !== true
    ) {
      return { path: path.split('.'), updates: undefined } as NormalizedResult;
    }

    const ast = this.parseCache[path] ?? parseBinding(path);
    this.parseCache[path] = ast;

    if (typeof ast !== 'object' || !ast?.status) {
      throw new TypeError(
        `Cannot normalize path "${path}": ${ast?.error ?? 'Unknown Error.'}`
      );
    }

    try {
      return resolveBindingAST(ast.path, resolveOptions, this.hooks);
    } catch (e: any) {
      throw new NestedError(`Cannot resolve binding: ${path}`, e);
    }
  }

  private getBindingForNormalizedResult(
    normalized: NormalizedResult
  ): BindingInstance {
    const normalizedStr = normalized.path.join('.');

    if (this.cache[normalizedStr]) {
      return this.cache[normalizedStr];
    }

    const created = new BindingInstance(
      normalizedStr === '' ? [] : normalized.path,
      this.parse
    );
    this.cache[normalizedStr] = created;

    return created;
  }

  public parse(
    rawBinding: BindingLike,
    overrides: Partial<BindingParserOptions> = {}
  ): BindingInstance {
    if (isBinding(rawBinding)) {
      return rawBinding;
    }

    const options = {
      ...this.parserOptions,
      ...overrides,
    };

    let updates: Record<string, any> = {};

    const joined = Array.isArray(rawBinding)
      ? rawBinding.join('.')
      : String(rawBinding);

    const normalizeConfig: ResolveBindingASTOptions = {
      getValue: (path: Array<string | number>) => {
        const normalized = this.normalizePath(path.join('.'), normalizeConfig);

        return options.get(this.getBindingForNormalizedResult(normalized));
      },
      evaluate: (exp) => {
        return options.evaluate(exp);
      },
      convertToPath: (path: any) => {
        if (path === undefined) {
          throw new Error(
            'Attempted to convert undefined value to binding path'
          );
        }

        if (
          typeof path !== 'string' &&
          typeof path !== 'number' &&
          typeof path !== 'boolean'
        ) {
          throw new Error(
            `Attempting to convert ${typeof path} to a binding path.`
          );
        }

        const normalized = this.normalizePath(String(path), normalizeConfig);

        if (normalized.updates) {
          updates = {
            ...updates,
            ...normalized.updates,
          };
        }

        const joinedNormalizedPath = normalized.path.join('.');

        if (joinedNormalizedPath === '') {
          throw new Error('Nested path resolved to an empty path');
        }

        return joinedNormalizedPath;
      },
    };

    const normalized = this.normalizePath(joined, normalizeConfig);

    if (normalized.updates) {
      updates = {
        ...updates,
        ...normalized.updates,
      };
    }

    const updateKeys = Object.keys(updates);

    if (!options.readOnly && updateKeys.length > 0) {
      const updateTransaction = updateKeys.map<[BindingInstance, any]>(
        (updatedBinding) => [
          this.parse(updatedBinding),
          updates[updatedBinding],
        ]
      );

      options.set(updateTransaction);
    }

    return this.getBindingForNormalizedResult(normalized);
  }
}
