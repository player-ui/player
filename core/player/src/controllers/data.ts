import { SyncHook, SyncWaterfallHook, SyncBailHook } from 'tapable-ts';
import { omit, removeAt } from 'timm';
import { dequal } from 'dequal';
import type { Logger } from '../logger';
import type { BindingParser, BindingLike } from '../binding';
import { BindingInstance } from '../binding';
import type {
  BatchSetTransaction,
  Updates,
  DataModelOptions,
  DataModelWithParser,
  DataPipeline,
  DataModelMiddleware,
} from '../data';
import { PipelinedDataModel, LocalModel } from '../data';
import type { RawSetTransaction } from '../types';

/** The orchestrator for player data */
export class DataController implements DataModelWithParser<DataModelOptions> {
  public hooks = {
    resolve: new SyncWaterfallHook(),
    resolveDataStages: new SyncWaterfallHook<[DataPipeline]>(),

    // On any set or get of an undefined value, redirect the value to be the default
    resolveDefaultValue: new SyncBailHook<[BindingInstance], any>(),

    onDelete: new SyncHook<[any]>(),

    onSet: new SyncHook<[BatchSetTransaction]>(),

    onGet: new SyncHook<[any, any]>(),

    onUpdate: new SyncHook<[Updates, DataModelOptions | undefined]>(),

    format: new SyncWaterfallHook<[any, BindingInstance]>(),

    deformat: new SyncWaterfallHook<[any, BindingInstance]>(),

    serialize: new SyncWaterfallHook<[any]>(),
  };

  private model?: PipelinedDataModel;
  private trash: Set<BindingInstance>;
  private pathResolver: BindingParser;
  private baseMiddleware: Array<DataModelMiddleware>;
  private logger?: Logger;

  constructor(
    model: Record<any, unknown> | undefined,
    options: {
      /** A means of parsing a raw binding to a Binding object */
      pathResolver: BindingParser;

      /** middleware to use. typically for validation */
      middleware?: Array<DataModelMiddleware>;

      /** A logger to use  */
      logger?: Logger;
    }
  ) {
    this.logger = options.logger;
    const middleware = options.middleware || [];
    this.baseMiddleware = [new LocalModel(model), ...middleware];

    this.trash = new Set();
    this.pathResolver = options.pathResolver;
  }

  public getModel(): PipelinedDataModel {
    if (!this.model) {
      const stages = this.hooks.resolveDataStages.call(this.baseMiddleware);
      const model = new PipelinedDataModel();
      model.setMiddleware(stages);
      this.model = model;
    }

    return this.model;
  }

  private resolveDataValue(
    binding: BindingInstance,
    value: any,
    deformat: boolean
  ) {
    if (deformat) {
      return this.hooks.deformat.call(value, binding);
    }

    return value;
  }

  public set(
    transaction: RawSetTransaction,
    options?: DataModelOptions
  ): Updates {
    let normalizedTransaction: BatchSetTransaction = [];

    if (Array.isArray(transaction)) {
      normalizedTransaction = transaction.map(([binding, value]) => {
        const parsed = this.pathResolver.parse(binding);

        return [
          parsed,
          this.resolveDataValue(parsed, value, Boolean(options?.formatted)),
        ];
      }) as BatchSetTransaction;
    } else {
      normalizedTransaction = Object.keys(transaction).map(
        (binding: string) => {
          const parsed = this.pathResolver.parse(binding);
          const val = transaction[binding];

          return [
            parsed,
            this.resolveDataValue(parsed, val, Boolean(options?.formatted)),
          ];
        }
      ) as BatchSetTransaction;
    }

    // Figure out what the base changes being applied are
    const setUpdates = normalizedTransaction.reduce<Updates>(
      (updates, [binding, newVal]) => {
        const oldVal = this.get(binding, { includeInvalid: true });

        const update = {
          binding,
          newValue: newVal,
          oldValue: oldVal,
        };

        if (dequal(oldVal, newVal)) {
          this.logger?.debug(
            `Skipping update for path: ${binding.asString()}. Value was unchanged: ${oldVal}`
          );
        } else {
          updates.push(update);

          this.logger?.debug(
            `Setting path: ${binding.asString()} from: ${oldVal} to: ${newVal}`
          );
        }

        return updates;
      },
      []
    );

    // Get the applied update
    const result = this.getModel().set(normalizedTransaction, options);

    // Add any extra bindings that were effected
    const setUpdateBindings = new Set(setUpdates.map((su) => su.binding));
    result.forEach((tr) => {
      if (
        !setUpdateBindings.has(tr.binding) &&
        (tr.force === true || !dequal(tr.oldValue, tr.newValue))
      ) {
        this.logger?.debug(
          `Path: ${tr.binding.asString()} was changed from: ${
            tr.oldValue
          } to: ${tr.newValue}`
        );
        setUpdates.push(tr);
      }
    });

    this.hooks.onSet.call(normalizedTransaction);

    if (setUpdates.length > 0) {
      this.hooks.onUpdate.call(setUpdates, options);
    }

    return result;
  }

  private resolve(binding: BindingLike, readOnly: boolean): BindingInstance {
    return Array.isArray(binding) || typeof binding === 'string'
      ? this.pathResolver.parse(binding, { readOnly })
      : binding;
  }

  public get(binding: BindingLike, options?: DataModelOptions) {
    const resolved =
      binding instanceof BindingInstance
        ? binding
        : this.resolve(binding, true);
    let result = this.getModel().get(resolved, options);

    if (result === undefined && !options?.ignoreDefaultValue) {
      const defaultVal = this.hooks.resolveDefaultValue.call(resolved);

      if (defaultVal !== result) {
        result = defaultVal;
      }
    }

    if (options?.formatted) {
      result = this.hooks.format.call(result, resolved);
    } else if (options?.formatted === false) {
      result = this.hooks.deformat.call(result, resolved);
    }

    this.hooks.onGet.call(binding, result);

    return result;
  }

  public delete(binding: BindingLike, options?: DataModelOptions) {
    if (
      typeof binding !== 'string' &&
      !Array.isArray(binding) &&
      !(binding instanceof BindingInstance)
    ) {
      throw new Error('Invalid arguments: delete expects a data path (string)');
    }

    const resolved =
      binding instanceof BindingInstance
        ? binding
        : this.resolve(binding, false);

    const parentBinding = resolved.parent();
    const property = resolved.key();
    const parentValue = this.get(parentBinding);

    const existedBeforeDelete =
      typeof parentValue === 'object' &&
      parentValue !== null &&
      Object.prototype.hasOwnProperty.call(parentValue, property);

    this.getModel().delete(resolved, options);

    if (existedBeforeDelete && !this.get(resolved)) {
      this.trash.add(resolved);
    }

    this.hooks.onDelete.call(resolved);
  }

  public serialize(): object {
    return this.hooks.serialize.call(this.get(''));
  }
}
