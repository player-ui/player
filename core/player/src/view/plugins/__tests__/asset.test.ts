import { describe, it, expect, beforeEach } from "vitest";
import { BindingParser } from "../../../binding";
import type { DataModelWithParser } from "../../../data";
import { LocalModel, withParser } from "../../../data";
import { ExpressionEvaluator } from "../../../expressions";
import { SchemaController } from "../../../schema";
import { Parser } from "../../parser";
import type { Options } from "../options";
import {
  MultiNodePlugin,
  AssetPlugin,
  ApplicabilityPlugin,
  TemplatePlugin,
  SwitchPlugin,
} from "..";

describe("asset", () => {
  let parser: Parser;

  beforeEach(() => {
    parser = new Parser();
    new AssetPlugin().applyParser(parser);
  });

  it("object", () => {
    expect(parser.parseObject({ asset: { type: "bar" } })).toMatchSnapshot();
  });

  it("applicability", () => {
    new ApplicabilityPlugin().applyParser(parser);
    new MultiNodePlugin().applyParser(parser);

    expect(
      parser.parseObject({
        asset: {
          values: [
            {
              applicability: "{{foo}}",
              value: "foo",
            },
            {
              value: "bar",
            },
          ],
        },
      }),
    ).toMatchSnapshot();
  });

  it("multi-node", () => {
    new MultiNodePlugin().applyParser(parser);

    expect(
      parser.parseObject({
        asset: {
          id: "foo",
          type: "collection",
          values: [
            {
              asset: {
                id: "value-1",
                type: "text",
                value: "First value in the collection",
              },
            },
          ],
        },
      }),
    ).toMatchSnapshot();
  });

  it("template", () => {
    const parseBinding = new BindingParser().parse;
    const model: DataModelWithParser = withParser(
      new LocalModel(),
      parseBinding,
    );
    const expressionEvaluator: ExpressionEvaluator = new ExpressionEvaluator({
      model,
    });
    const options: Options = {
      evaluate: expressionEvaluator.evaluate,
      schema: new SchemaController(),
      data: {
        format: (binding, val) => val,
        formatValue: (val) => val,
        model,
      },
    };
    new TemplatePlugin(options).applyParser(parser);

    const petNames = ["Ginger", "Daisy", "Afra"];
    model.set([["foo.bar", petNames]]);

    expect(
      parser.parseObject({
        asset: {
          id: "foo",
          type: "collection",
          template: [
            {
              data: "foo.bar",
              output: "values",
              value: {
                value: "{{foo.bar._index_}}",
              },
            },
          ],
        },
      }),
    ).toMatchSnapshot();
  });

  it("switch", () => {
    new SwitchPlugin({
      evaluate: () => {
        return true;
      },
    } as any).applyParser(parser);

    expect(
      parser.parseObject({
        id: "toughView",
        type: "view",
        title: {
          staticSwitch: [
            {
              case: "'true'",
              asset: {
                id: "test",
                type: "text",
                value: "test-text.",
              },
            },
          ],
        },
      }),
    ).toMatchSnapshot();
  });
});
