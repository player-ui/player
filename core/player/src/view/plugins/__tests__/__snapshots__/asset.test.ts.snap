// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`asset > applicability 1`] = `
{
  "children": [
    {
      "path": [
        "asset",
      ],
      "value": {
        "children": [
          {
            "path": [
              "values",
            ],
            "value": {
              "override": true,
              "parent": [Circular],
              "type": "multi-node",
              "values": [
                {
                  "expression": "{{foo}}",
                  "parent": [Circular],
                  "type": "applicability",
                  "value": {
                    "parent": [Circular],
                    "type": "value",
                    "value": {
                      "value": "foo",
                    },
                  },
                },
                {
                  "parent": [Circular],
                  "type": "value",
                  "value": {
                    "value": "bar",
                  },
                },
              ],
            },
          },
        ],
        "parent": [Circular],
        "type": "asset",
        "value": undefined,
      },
    },
  ],
  "type": "value",
  "value": undefined,
}
`;

exports[`asset > multi-node 1`] = `
{
  "children": [
    {
      "path": [
        "asset",
      ],
      "value": {
        "children": [
          {
            "path": [
              "values",
            ],
            "value": {
              "override": true,
              "parent": [Circular],
              "type": "multi-node",
              "values": [
                {
                  "children": [
                    {
                      "path": [
                        "asset",
                      ],
                      "value": {
                        "parent": [Circular],
                        "type": "asset",
                        "value": {
                          "id": "value-1",
                          "type": "text",
                          "value": "First value in the collection",
                        },
                      },
                    },
                  ],
                  "parent": [Circular],
                  "type": "value",
                  "value": undefined,
                },
              ],
            },
          },
        ],
        "parent": [Circular],
        "type": "asset",
        "value": {
          "id": "foo",
          "type": "collection",
        },
      },
    },
  ],
  "type": "value",
  "value": undefined,
}
`;

exports[`asset > object 1`] = `
{
  "children": [
    {
      "path": [
        "asset",
      ],
      "value": {
        "parent": [Circular],
        "type": "asset",
        "value": {
          "type": "bar",
        },
      },
    },
  ],
  "type": "value",
  "value": undefined,
}
`;

exports[`asset > switch 1`] = `
{
  "children": [
    {
      "path": [
        "title",
        "asset",
      ],
      "value": {
        "parent": [Circular],
        "type": "asset",
        "value": {
          "id": "test",
          "type": "text",
          "value": "test-text.",
        },
      },
    },
  ],
  "type": "value",
  "value": {
    "id": "toughView",
    "type": "view",
  },
}
`;

exports[`asset > template 1`] = `
{
  "children": [
    {
      "path": [
        "asset",
      ],
      "value": {
        "children": [
          {
            "path": [
              "values",
            ],
            "value": {
              "override": false,
              "parent": [Circular],
              "type": "multi-node",
              "values": [
                {
                  "parent": [Circular],
                  "type": "value",
                  "value": {
                    "value": "{{foo.bar.0}}",
                  },
                },
                {
                  "parent": [Circular],
                  "type": "value",
                  "value": {
                    "value": "{{foo.bar.1}}",
                  },
                },
                {
                  "parent": [Circular],
                  "type": "value",
                  "value": {
                    "value": "{{foo.bar.2}}",
                  },
                },
              ],
            },
          },
        ],
        "parent": [Circular],
        "type": "asset",
        "value": {
          "id": "foo",
          "type": "collection",
        },
      },
    },
  ],
  "type": "value",
  "value": undefined,
}
`;
